master/buildbot/master.py

```diff
@@ -334,53 +334,29 @@ def sendBuildbotNetUsageData(self):
 
     @defer.inlineCallbacks
     def stopService(self):
-        if self.masterid is not None:
-            yield self.data.updates.masterStopped(
-                name=self.name, masterid=self.masterid)
-        if self.running:
-            yield self.botmaster.cleanShutdown(quickMode=True, stopReactor=False)
-            yield service.AsyncMultiService.stopService(self)
-
-        log.msg("BuildMaster is stopped")
-        self._master_initialized = False
+        try:
+            yield self.initLock.acquire()
+            if self.masterid is not None:
+                yield self.data.updates.masterStopped(
+                    name=self.name, masterid=self.masterid)
+            if self.running:
+                yield self.botmaster.cleanShutdown(
+                    quickMode=True, stopReactor=False)
+                yield service.AsyncMultiService.stopService(self)
+
+            log.msg("BuildMaster is stopped")
+            self._master_initialized = False
+        finally:
+            yield self.initLock.release()
 
+    @defer.inlineCallbacks
     def reconfig(self):
-        # this method wraps doConfig, ensuring it is only ever called once at
-        # a time, and alerting the user if the reconfig takes too long
-        if self.reconfig_active:
-            log.msg("reconfig already active; will reconfig again after")
-            self.reconfig_requested = True
-            return
-
-        self.reconfig_active = self.reactor.seconds()
-        metrics.MetricCountEvent.log("loaded_config", 1)
-
-        # notify every 10 seconds that the reconfig is still going on, although
```

Comment:
I think you should still keep the long reconfig warning timer.

Also, if you send 10 SIGHUP in a row, you'll get 10 reconfig, while the previous implementation, only 2